server/punishments.ts

```diff
@@ -861,17 +864,22 @@ export const Punishments = new class {
 		}
 		return success;
 	}
-	automaticUnlock(user: User, punishedID: ID | PunishType) {
-		if (user.locked === punishedID) {
-			Punishments.unlock(user.id);
-		}
-	}
 	/**
-	 * Validates the length and sets the punishment timer for a user.
+	 * Sets the punishment timer for a user,
+	 * to either MAX_PUNISHMENT_TIMER_LENGTH or the amount of time left on the punishment.
+	 * It also expires a punishment if the time is up.
 	 */
-	setPunishmentTimer(user: User, expireTime: number, punishedID: ID | PunishType) {
-		if (expireTime === 0 || expireTime === Infinity || expireTime > Chat.MAX_TIMEOUT_DURATION) return;
-		user.punishmentTimer = setTimeout(Punishments.automaticUnlock, expireTime - Date.now(), user, punishedID);
+	checkPunishmentTime(user: User, punishment: Punishment) {
+		const [, id, expireTime] = punishment;
+
+		if (expireTime < Date.now()) {
+			if (user.locked === id) Punishments.unlock(user.id);
+		} else {
+			const length = Math.min(expireTime - Date.now(), MAX_PUNISHMENT_TIMER_LENGTH);
+			user.punishmentTimer = setTimeout(() => {
+				Punishments.checkPunishmentTime(user, punishment);
+			}, length);
+		}
```

Comment:
```suggestion
		const timeLeft = expireTime - Date.now();
		if (timeLeft <= 1) {
			if (user.locked === id) Punishments.unlock(user.id);
			return;
		}
		const waitTime = Math.min(timeLeft, MAX_PUNISHMENT_TIMER_LENGTH);
		user.punishmentTimer = setTimeout(() => {
			Punishments.checkPunishmentTime(user, punishment);
		}, waitTime);
```

Reply:
Mostly, this is about not calling `Date.now()` multiple times. In this case it's not a big deal, but it's just a good habit to get into. Naming variables `timeLeft` and `waitTime` I think improves readability by a lot.